### PR TITLE
GPU Acceleration achieved!

### LINKED ISSUE #260: Blur module producing a black 'border' at the top and bottom
## Example Image
Example Image
## Example Image Blurred
Example Image Blurred

Is this the intended behaviour of the algorithm or is this a bug?

### LINKED ISSUE #849: Accelerate Convolution with GPU.js
### Kernel Convolution
This is the process of convolving a small matrix(called a kernel) of size 3\*3 or 5\*5 or any odd\*odd size over a larger matrix(of pixels) to get values of pixels based on the adjacent pixels. This technique is used in Gaussian blur, mean blur, filters, edge detection and our general purpose convolution module.

### Performance
This technique is very slow as it has to do a lot of processing of pixels and lots of math operations.

### GPGPU
GPGPU stands for general purpose programming on GPU. This is a way to write functions which dispatch many parallel operations to the GPU to accelerate the process. It is mainly used for matrix multiplication, matrix transformation and convolution.

### Kernel
A kernel, not to be confused with the convolved matrix, is a fancy word for a function run on the GPU.

### GPU.js
[GPU.js](https://github.com/gpujs/gpu.js) is a library which compiles simple js functions into a GPGPU language like GLSL and runs all the calculations in parallel on a GPU. If a GPU is not available, it will fallback to normal js on the CPU. This can be used not only for convolution but also for modules which manipulate many pixels at once. Maybe even for the pixelManipulation API.

****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site.

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### LINKED ISSUE #998: Bugs in blur module
The amount of gaussian blur is not changing even after changing the intensity value.

blur

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site.

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by HarshKhandeparkar**
Fixes #849
Fixes #998
Fixes #260 

#### Changes
1. Blur module produces an actual Gaussian kernel based on the given sigma.
blur-better

#### Added
1. Blur module is GPU accelerated.
2. We have a new util module `src/modules/_nomodule/gpuUtils.js`. It exports `convolve` for convolution and `compute` for general computing on a GPU using JS!
3. Edge Detect Uses our blur module instead of using the `ndarray-gaussian-blur` package. Edge Detect was bottlenecked because of `ndarray-gaussian-blur`.

#### Better
1. Edge Detect runs much, much faster!
edge-detect-faster
##### Here is a GIF of the old EdgeDetect
old-edge-det

2. Blur no longer produces a black border! New blur module takes care of the black border by adding some padding to the pixels.

3. *Convolution* module runs at **60fps**
new-conv
This is the old one for comparison
old-conv

#### Things not done
1. Edge Detect convolution was not done on a GPU: Edge Detect convolution was already blazing fast, it was the blur module which consumed time. GPU computing requires compilation etc., overall, for edge detect which has a 3x3 kernel was actually slower on a GPU than on a CPU. Blur which currently has a 5x5 kernel is benefitted from a GPU. This kernel size can be easily bumped in the future for a higher quality of blur.
2. I had mentioned that changePixel could have been done on a GPU, turns out, it isn't beneficial because of the reasons listed above.

#### Best Part: It all works in **NodeJS**
Screenshot from 2019-04-23 01-42-55

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/is-reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by HarshKhandeparkar** [comment]
@tech4GT @VibhorCodecianGupta @jywarren @publiclab/is-reviewers big PR!

**Comment by HarshKhandeparkar** [comment]
@tech4GT @VibhorCodecianGupta @jywarren this was working on my PC. Travis doesn't have the package `libxi-dev` installed by default. I tried to installed it but it doesn't seem to be working. Could you please help me? I don't know much about travis.

**Comment by HarshKhandeparkar** [comment]
Ok the package was successfully installed. Now it is showing a different error. I cannot track it. It says that it cannot recognise `...` which I think is a spread operator.

**Comment by HarshKhandeparkar** [comment]
I didn't even add something like that.

**Comment by rexagod** [comment]
```sh
  total:     14
  passing:   14
  duration:  6.6s


npm ERR! Test failed.  See above for more details.
```
Notice that this happens on local as well, @HarshKhandeparkar.

**Comment by rexagod** [comment]
What's odd is the build errored even though all tests pass. Probably need to figure out which file in your `npm test` script is causing this.

**Comment by rexagod** [comment]
I'll take a look, but do share any ideas you might have that can help us narrow this down.

**Comment by HarshKhandeparkar** [comment]
I found the file
It happens in gpu.js where a spread operator is used. The weird thing is that it recognizes a spread operator which is used in IS in edge detect module

**Comment by HarshKhandeparkar** [comment]
We will probably have to use babel-tape-runner

**Comment by HarshKhandeparkar** [comment]
It seems that browserify doesn't support object spread.

**Comment by HarshKhandeparkar** [comment]
I have used babelify to transform the object rest spread and the tests pass(at least locally)

**Comment by HarshKhandeparkar** [comment]
I have used babelify to transform the object rest spread and the tests pass(at least locally)

**Comment by HarshKhandeparkar** [comment]
Ok, I fixed one error but now a new one has popped up. The tests pass locally on my pc but fail in Travis.

**Comment by HarshKhandeparkar** [comment]
Maybe it is failing because there isn't a GPU on the travis servers.

**Comment by HarshKhandeparkar** [comment]
It would normally fall back to cpu mode in absence of a gpu but this one is a pre-release so it may be buggy. In any case, a normal day-to-day user should get no errors on a pc but I would still like the others to clone the branch and run the tests. Thanks.
cc @publiclab/is-reviewers @rexagod

**Comment by HarshKhandeparkar** [comment]
@robertleeplummerjr please have a look at the CI.

**Comment by robertleeplummerjr** [comment]
CI issues _should_ be resolved in https://github.com/gpujs/gpu.js/releases/tag/2.0.0-rc.8.  Please upgrade and let me know.

**Comment by HarshKhandeparkar** [comment]
Thanks a lot @robertleeplummerjr

**Comment by HarshKhandeparkar** [comment]
@robertleeplummerjr didn't fix it. :confused:

**Comment by rexagod** [comment]
@HarshKhandeparkar I'm going to fork your IS repo and run some custom travis scripts to narrow this down.

**Comment by HarshKhandeparkar** [comment]
We got the error.

**Comment by HarshKhandeparkar** [comment]
Thanks for your help.

**Comment by HarshKhandeparkar** [comment]
Do the tests fail locally? What OS?

**Comment by rexagod** [comment]
One thing though, was `WebGL` used in anywhere in this PR? Also, I just need to wrap few things up at the LDI repository, then I can focus on this, it won't take long!

**Comment by HarshKhandeparkar** [comment]
> One thing though, was WebGL used in anywhere in this PR?

Everywhere in this PR!

**Comment by rexagod** [comment]
Yup, check the last line of this log, although all tests pass.

```sh
 % npm test

> image-sequencer@3.3.0 test /mnt/c/Users/Rexagod/Desktop/is
> TEST=true istanbul cover tape test/core/*.js test/core/ui/user-interface.js test/core/modules/canvas-resize.js test/core/modules/QR.js test/core/modules/crop.js | tap-spec; browserify --global-transform [ babelify --plugins [ @babel/plugin-syntax-object-rest-spread @babel/plugin-proposal-object-rest-spread ] ] test/core/modules/image-sequencer.js test/core/modules/chain.js test/core/modules/meta-modules.js test/core/modules/replace.js test/core/modules/import-export.js test/core/modules/run.js test/core/modules/dynamic-imports.js test/core/util/parse-input.js test/core/modules/benchmark.js| tape-run --render="tap-spec"


  Output directory is correctly generated

    ✔ Access the created dir

  Notify function works for all three UIs

    ✔ Default notify() produces correct output
    ✔ Html UI contains notify function

  Load canvas-resize module

    ✔ Canvas resize module is getting loaded

  Check Options

    ✔ Options are correct
    ✔ Options are correct

  canvas-resize module works correctly

    ✔ should be equal
    ✔ should be equal

  Load add QR module

    ✔ Add-QR module is getting loaded.

  Check Options

    ✔ Options are correct

  Add QR module works correctly

    ✔ should be equal

  Crop module loads correctly

    ✔ Crop module loaded

  Crop module loads with correct options

    ✔ options loaded correctly

  Crop module works correctly

    ✔ should be equal
=============================================================================
Writing coverage object [/mnt/c/Users/Rexagod/Desktop/is/coverage/coverage.json]
Writing coverage reports at [/mnt/c/Users/Rexagod/Desktop/is/coverage]
=============================================================================
    =============================== Coverage summary ===============================
    Statements   : 35.11% ( 697/1985 )
    Branches     : 27.48% ( 177/644 )
    Functions    : 20.63% ( 72/349 )
    Lines        : 35.75% ( 686/1919 )
    ================================================================================


  total:     14
  passing:   14
  duration:  1m 14.9s


[BABEL] Note: The code generator has deoptimised the styling of /mnt/c/Users/Rexagod/Desktop/is/node_modules/lodash/lodash.js as it exceeds the max of 500KB.
npm ERR! Test failed.  See above for more details.
```

**Comment by rexagod** [comment]
Do you have any ideas you'd like to try out first, or should I start with the scripts?

**Comment by HarshKhandeparkar** [comment]
> npm ERR! Test failed.  See above for more details.

It doesn't show the error :confused:

**Comment by rexagod** [comment]
Fingers crossed!

**Comment by HarshKhandeparkar** [comment]
@robertleeplummerjr it failed but it has a different error.

**Comment by robertleeplummerjr** [comment]
Reproduced, and fixing.

**Comment by HarshKhandeparkar** [comment]
Thanks a lot!

File: src/modules/Blur/Blur.js [hunk 1]
```diff
@@ -1,84 +1,136 @@
...
-                    acc[1] += (oldpix.get(neighboutPos[a][b][0], neighboutPos[a][b][1], 1) * kernel[a][b]);
-                    acc[2] += (oldpix.get(neighboutPos[a][b][0], neighboutPos[a][b][1], 2) * kernel[a][b]);
-                    acc[3] += (oldpix.get(neighboutPos[a][b][0], neighboutPos[a][b][1], 3) * kernel[a][b]);
-                }
-            }
-            pixels.set(i, j, 0, acc[0]);
-            pixels.set(i, j, 1, acc[1]);
-            pixels.set(i, j, 2, acc[2]);
-        }
+  let kernel = kernelGenerator(blur);
```
Comment (jywarren) [inline src/modules/Blur/Blur.js]: This is awesome. Should we think about keeping the original as a fallback for environments where the new version wouldn't work? Are there any? Like a device with no GPU?

Thanks!

Comment (HarshKhandeparkar) [reply src/modules/Blur/Blur.js]: @jywarren GPU.js has an internal fallback but it is not working which is the bug that Robert is fixing. It will mostly get fixed in the next rc(which will release today).

**Comment by rexagod** [comment]
@HarshKhandeparkar I did have an idea about this recently, which was inspired by the discussion [here](https://github.com/mrdoob/three.js/issues/4927). I sent a request to https://get.webgl.org/ from within travis itself to check whether the machine was capable of running WebGL and/or if the module was maybe disabled by default, but the results were positive, i.e., there was no problem with the code itself, but the dependency.

Let's hope the next `rc` will be stable and solve this as well!

**Comment by HarshKhandeparkar** [comment]
Yes @rexagod the problem isn't WebGL, problem is that Travis doesn't have a GPU. In a normal scenario, GPU.js will fallback to pure javascript but it didn't. That bug was also fixed. Now there is a single integer literal which is causing problems. Robert has assured me that it will get fixed in the next rc but there is no guarantee that it will be 100% stable as it is an rc after all. The rc will probably be published later today. Thanks for your help.

**Comment by jywarren** [comment]
Ok awesome!!! Thanks for the clarification!!

**Comment by HarshKhandeparkar** [comment]
Np

**Comment by HarshKhandeparkar** [comment]
Travis won't work until I resolve conflicts 😔. I will resolve them later today.

**Comment by robertleeplummerjr** [comment]
I'll look into latest issue later today.  It is slightly different than the previous.

**Comment by HarshKhandeparkar** [comment]
@jywarren tests pass! could you please help me with codeclimate?

**Comment by Divy123** [comment]
@HarshKhandeparkar  do you mean code coverage?

**Comment by HarshKhandeparkar** [comment]
no, the CI

**Comment by Divy123** [comment]
You can close and reopen the PR, sometimes it happens.

**Comment by HarshKhandeparkar** [comment]
really?

**Comment by HarshKhandeparkar** [comment]
I meant [this](https://codeclimate.com/github/publiclab/image-sequencer/pull/1038)

**Comment by Divy123** [comment]
Sorry but I thought you were talking about Travis CI...
Can you explain a bit what help do you need here?

**Comment by HarshKhandeparkar** [comment]
@Divy123 I don't even know when it was added. Neither do I know how to fix it (not that I know what the errors mean). Could anyone help me? @publiclab/is-reviewers

**Comment by jywarren** [comment]
It makes code refactoring recommendations... we just turned it on, but we can "accept" to decline the recommendations. This looks good, but @Divy123 would you mind giving it a review? Thank you!!!

File: src/modules/Blur/Blur.js [hunk 2]
```diff
@@ -1,84 +1,136 @@
...
+
+  // let oldpix = require('lodash').cloneDeep(pixels);
```
Comment (jywarren) [inline src/modules/Blur/Blur.js]: Shall we remove these sections?

File: test/core/util/gpuUtils.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
+const test = require('tape');
```
Comment (jywarren) [inline test/core/util/gpuUtils.js]: ooooo very nice!!!!

**Comment by Divy123** [comment]
@HarshKhandeparkar yes code climate was already there but PR integration is done today itself by @jywarren. I think it is utmost for the codebase. I had raised an issue earlier and we ended up with this .

Jeff has approved them here.

File: src/modules/Blur/Blur.js (hunk 2 above)
Comment (HarshKhandeparkar) [reply src/modules/Blur/Blur.js]: we can, but if we need to fall back to the old version at any time in the future, we can use this. They are comments so they do not affect performance. I can still remove them though.

**Comment by Divy123** [comment]
Ya sure @jywarren!!

**Comment by HarshKhandeparkar** [comment]
Fixing conflicts in a min...

**Comment by HarshKhandeparkar** [comment]
fixed(I guess)

**Comment by HarshKhandeparkar** [comment]
Oh, thanks a lot for approving the CI @jywarren.

**Comment by HarshKhandeparkar** [comment]
@jywarren please review this.

**Comment by HarshKhandeparkar** [comment]
@publiclab/is-reviewers please review :) I don't have much time and I want to get this merged. Thanks

**Comment by HarshKhandeparkar** [comment]
@jywarren why do I have merge rights?

**Comment by Divy123** [comment]
@HarshKhandeparkar I will also be reviewing this in an hour. Actually my semester lab exams have started so a bit late. Will do this in an hour.

**Comment by HarshKhandeparkar** [comment]
@Divy123 thanks

File: src/modules/Blur/Module.js [hunk 4]
```diff
@@ -15,7 +15,7 @@ module.exports = function Blur(options, UI) {
         var step = this;
 
         function extraManipulation(pixels) {
-            pixels = require('./Blur')(pixels, options.blur);
+            pixels = require('./Blur')(pixels, parseFloat(options.blur));
```
Comment (Divy123) [inline src/modules/Blur/Module.js]: @HarshKhandeparkar I think parseFloat can be done seperately so that its clear.

File: src/modules/Convolution/Convolution.js [hunk 5]
```diff
@@ -1,71 +1,98 @@
 var _ = require('lodash');
-module.exports = exports = function(pixels, constantFactor, kernelValues) {
-    let kernel = kernelGenerator(constantFactor, kernelValues), oldpix = _.cloneDeep(pixels);
-    kernel = flipKernel(kernel);
+module.exports = exports = function(pixels, constantFactor, kernelValues, texMode) {
+  let kernel = kernelGenerator(constantFactor, kernelValues);
+  
+  // const  oldpix = _.cloneDeep(pixels);
```
Comment (Divy123) [inline src/modules/Convolution/Convolution.js]: Any reasons for this much comments?

File: src/modules/Blur/Module.js (hunk 4 above)
Comment (HarshKhandeparkar) [reply src/modules/Blur/Module.js]: Do you mean like `options.blur = parseFloat(options.blur)`?

File: src/modules/Blur/Blur.js [hunk 6]
```diff
@@ -1,84 +1,136 @@
...
+      pixels.set(x, y, 2, Math.max(0, Math.min(conPix[2][y][x], 255)))
     }
-    function getNeighbouringPixelPositions(pixelPosition) {
-        let x = pixelPosition[0], y = pixelPosition[1], result = [];
+  }
+
+  return pixels;
+
+  //Generates a 5x5 Gaussian kernel
+  function kernelGenerator(sigma = 1) {
```
Comment (Divy123) [inline src/modules/Blur/Blur.js]: Nice use of defaults here

**Comment by Divy123** [comment]
Yes this way it will be more clear.

File: src/modules/Convolution/Convolution.js (hunk 5 above)
Comment (HarshKhandeparkar) [reply src/modules/Convolution/Convolution.js]: > if we need to fall back to the old version at any time in the future, we can use this. They are comments so they do not affect performance. I can still remove them though.

**Comment by Divy123** [comment]
In that case I think we should maintain a separate file because that would
make this one too heavy with some redundant data.
Maintaining two files can be better.
I mean we can keep the older version can be kept in one file and the newer
in another.
This will be easier to maintain.
Please tell me if I understood incorrectly.

**Comment by HarshKhandeparkar** [comment]
Or I can just delete the comments if not needed. Saving in a different file doesn't really do much. Should I delete them?

**Comment by Divy123** [comment]
You can do so if they are not really required.

Or you can collect them at the bottom by providing appropriate references.

**Comment by HarshKhandeparkar** [comment]
@Divy123 committed the changes you had requested.

**Comment by jywarren** [comment]
Hi @HarshKhandeparkar - would you be able to add a simple test for Convolution and Edge Detect, to keep our coverage from falling? Neither of those two modules have tests, and CodeCov is pointing it out:

https://github.com/publiclab/image-sequencer/pull/1038#issuecomment-487694890

What do you think? With that, i'd be very happy to merge this!

**Comment by jywarren** [comment]
Also, @HarshKhandeparkar this is so cool. 🎉 

Do you think it would be worth refactoring Webgl-distort to use this? would it be faster in a server-node environment than starting up a puppeteer instance? https://github.com/tech4GT/image-sequencer/blob/a3f1b016dacfee38b55966cab63c917094dd3d1e/src/modules/WebglDistort/Module.js#L154

**Comment by HarshKhandeparkar** [comment]
> Hi @HarshKhandeparkar - would you be able to add a simple test for Convolution and Edge Detect, to keep our coverage from falling? Neither of those two modules have tests, and CodeCov is pointing it out:
> 
> [#1038 (comment)](https://github.com/publiclab/image-sequencer/pull/1038#issuecomment-487694890)
> 
> What do you think? With that, i'd be very happy to merge this!

I want to add tests for modules, all the remaining ones in a followup as soon as this is merged. Is this ok?

**Comment by HarshKhandeparkar** [comment]
> Also, @HarshKhandeparkar this is so cool. tada
> 
> Do you think it would be worth refactoring Webgl-distort to use this? would it be faster in a server-node environment than starting up a puppeteer instance? https://github.com/tech4GT/image-sequencer/blob/a3f1b016dacfee38b55966cab63c917094dd3d1e/src/modules/WebglDistort/Module.js#L154

Maybe, but we will have to write the distort code in JS rather than GLSL and also the compilation takes a while, a long while. The bigger the kernel, longer the time. Also, if the output image can be bigger than the input image, handling that will also require more computation. The decision is up to the community. I have just added a General Purpose GPU computation utility method called `compute` in `src/modules/_nomodule/gpuUtils.js`. That one can be used if we ever decide to do so.

**Comment by jywarren** [comment]
Hi, @HarshKhandeparkar I would really like to see the tests associated within a single PR, and our coverage monitoring would appreciate it too! We can wait on this one for the time being if you are busy, but I think we should start accepting tests + new code together as a matter of policy, in order to really ratchet up our coverage and stability. I hope you understand! ❤️ Thank you for your fantastic work!

**Comment by HarshKhandeparkar** [comment]
This PR concerns GPU acceleration and that is tested. The modules are still
the same so I want to do that in a separate PR as they are not related to
this one. They just use a different tool and the tool is tested.

**Comment by jywarren** [comment]
I hear you, @HarshKhandeparkar, and I really empathize with your desire to close this very large PR up. I do want to note that you've changed the internals of those modules, and I worry a little bit that without tests, we can't see that they are working in Travis. Although, without tests before, we couldn't see that they were working before either! 

So, because this is a gradual process, I'm going to merge this, and I think we'll all feel great about it 🙌 

However, I think we will be getting more and more committed to always asking for tests with any module changes at all as time goes on. I'd LOVE it if you would consider filing a follow-up quickly to help improve our coverage over these modules specifically! Thank you so much! ❤️ 👍 🎉

**Comment by HarshKhandeparkar** [comment]
@jywarren thanks a lot. I will open a follow-up PR as soon as I can. If it is possible today then today or tomorrow.
